jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java

```diff
@@ -466,15 +377,84 @@ private void clearBuckets(RetainedBucket[] buckets)
         }
     }
 
-    @Override
-    public void dump(Appendable out, String indent) throws IOException
+    private void releaseExcessMemory(boolean direct, int capacity)
     {
-        Dumpable.dumpObjects(
-            out,
-            indent,
-            this,
-            DumpableCollection.fromArray("direct", _direct),
-            DumpableCollection.fromArray("indirect", _indirect));
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (maxMemory > 0)
+        {
+            long excess = getMemory(direct) - maxMemory;
```

Comment:
This is a full iteration over buckets with a full iteration over each concurrent pool entry for every pool miss.       This is not different to my PRs always checking memory in the retain method.

However, you do not check max memory for a normal release, and it certainly can be exceeded (especially with the strange impl of getMemory).

Reply:
It's worse than your impl, and the memory accounting is quite strange indeed.